internal/controller/atlasorgsettings/handler_test.go

```diff
@@ -0,0 +1,711 @@
+			} else {
+				require.NoError(t, err)
+				assert.NotNil(t, reconcileCtx)
+				if tt.expectService {
+					assert.NotNil(t, reconcileCtx.svc)
+				}
+				assert.Equal(t, tt.input, reconcileCtx.aos)
+			}
+		})
+	}
+}
+
+func TestUpsert(t *testing.T) {
+	scheme := runtime.NewScheme()
+	require.NoError(t, corev1.AddToScheme(scheme))
+	require.NoError(t, akov2.AddToScheme(scheme))
+	ctx := context.Background()
+
+	tests := []struct {
+		name           string
+		currentState   state.ResourceState
+		nextState      state.ResourceState
+		provider       atlas.Provider
+		serviceBuilder func(*atlas.ClientSet) atlasorgsettings.AtlasOrgSettingsService
+		input          *akov2.AtlasOrgSettings
+		objects        []client.Object
+		want           ctrlstate.Result
+		wantErr        string
+	}{
+		{
+			name:         "successful upsert with settings different - update needed",
+			currentState: state.StateInitial,
+			nextState:    state.StateCreated,
+			provider:     createSuccessfulProvider(),
+			serviceBuilder: createServiceBuilder(t,
+				&atlasorgsettings.AtlasOrgSettings{
+					AtlasOrgSettingsSpec: akov2.AtlasOrgSettingsSpec{
+						OrgID:                 fakeOrgID,
+						ApiAccessListRequired: pointer.MakePtr(false), // Different from sample
+					},
+				}, nil,
+				&atlasorgsettings.AtlasOrgSettings{
+					AtlasOrgSettingsSpec: akov2.AtlasOrgSettingsSpec{
+						OrgID:                 fakeOrgID,
+						ApiAccessListRequired: pointer.MakePtr(true),
+					},
+				}, nil, true),
+			input:   &sampleAtlasOrgSettings,
+			objects: []client.Object{&fakeAtlasSecret},
+			want: ctrlstate.Result{
+				NextState: "Created",
+				StateMsg:  "Updated.",
+			},
+		},
+		{
+			name:         "successful upsert with identical settings - no update needed",
+			currentState: state.StateInitial,
+			nextState:    state.StateCreated,
+			provider:     createSuccessfulProvider(),
+			serviceBuilder: createServiceBuilder(t,
+				&atlasorgsettings.AtlasOrgSettings{
+					AtlasOrgSettingsSpec: akov2.AtlasOrgSettingsSpec{
+						OrgID:                                  fakeOrgID,
+						ApiAccessListRequired:                  sampleAtlasOrgSettings.Spec.ApiAccessListRequired,
+						GenAIFeaturesEnabled:                   sampleAtlasOrgSettings.Spec.GenAIFeaturesEnabled,
+						MaxServiceAccountSecretValidityInHours: sampleAtlasOrgSettings.Spec.MaxServiceAccountSecretValidityInHours,
+						MultiFactorAuthRequired:                sampleAtlasOrgSettings.Spec.MultiFactorAuthRequired,
+						RestrictEmployeeAccess:                 sampleAtlasOrgSettings.Spec.RestrictEmployeeAccess,
+						SecurityContact:                        sampleAtlasOrgSettings.Spec.SecurityContact,
+						StreamsCrossGroupEnabled:               sampleAtlasOrgSettings.Spec.StreamsCrossGroupEnabled,
+					},
+				}, nil, nil, nil, false), // No update expected
+			input:   &sampleAtlasOrgSettings,
+			objects: []client.Object{&fakeAtlasSecret},
+			want: ctrlstate.Result{
+				NextState: "Created",
```

Comment:
I am not sure this resource should have the concept of created, it seems to me it is more like an IP access list kind of thing:
https://docs.google.com/presentation/d/1j1A7uF72C672Ct766O-ozj_BAwBFpRazdtTxb5o6Oag/edit?slide=id.g35221247c10_0_81#slide=id.g35221247c10_0_81

Only relevant state machines states are:
**Initial**
**Updated**
**Import Requested**
**Imported**
**DeletionRequested** (which is only for the Kubernetes side of things, as it does not delete it from Atlas, but rather un manages it as the settings will exist so long as the org does)
**Deleted**

Reply:
We do have this state: https://github.com/mongodb/mongodb-atlas-kubernetes/blob/4cdcc17fbb725bd3eab17bc5925ac91d3df9c94d/pkg/state/state.go#L38. But I'm just thinking, since OrgSettings are already there, they can only be "Updated".